test/run-drun/stable-log.mo

```diff
@@ -0,0 +1,73 @@
+import Prim "mo:⛔";
+import StableMemory "stable-mem/StableMemory";
+
+actor {
+
+  func ensure(offset : Nat32) {
+    let pages = (offset + 65536) >> 16;
+      if (pages > StableMemory.size()) {
+        let oldsize = StableMemory.grow(pages - StableMemory.size());
+        assert (oldsize != 0xFFFF);
+     };
+  };
+
+  stable var base : Nat32 = 0;
+
+  public func log(t : Text) {
+    let blob = Prim.encodeUtf8(t);
+    let size = Prim.natToNat32(blob.size());
+    let next = base + 4 + size;
+    ensure(next);
+    StableMemory.storeNat32(base, size);
+    StableMemory.storeBlob(base + 4, blob);
+    base := next;
+  };
+
+  public query func read(offset : Nat32) : async ?(Text, nextOffset : Nat32) {
```

Comment:
This strikes me as unsafe API, since the offset might be completely unaligned. Better be private? (For a public method, I'd rather use an index, or some abstract iteration API).

Reply:
Right, but how do you enforce abstraction? We don't have any and, for a query, you can't maintain any state on the server.

Reply:
@rossberg can you think of any solutions?

Reply:
It doesn't matter for this test, I'd just make it private.

For a real-world lib, it would probably need an indirection through an index. The simplest way would be a stable var with a map entry# -> address. But that would be kind of cheating, since then upgrade is O(N). To avoid that, you'd have to fold the map's tree into the stable memory as well. I could imagine ways of doing that, e.g., with index blocks of a fixed size are followed by their actual messages. Each such block can have two successor blocks When the current block fills up, you find the next unused successor in a breadth-first traversal and append a new block. I think this could be done with amortised O(1) append and O(log N) index lookup. Something like that, just thinking aloud...

Reply:
Queries can't be private, and that would defeat the point of having a quickly readable log (from the outside at least).
Maybe I could make it a stack and let you query the last n entries...

Reply:
I want the example to be simple.

Reply:
> If you want simple, maybe have read take an index and iterate through the log, and accept that it's O(N) (maybe have the index reversed, as you suggest)?

I believe that is a reasonable cost for the simplicity that @crusso understandably wants here.  Wanting more seems unreasonable without having a more complex representation, which would undermine the value of the example.

> I'd definitely not showcase an unsafe API design that leaks (and accepts!) verbatim backing store addresses.

> For a real-world lib, it would probably need an indirection through an index. 

Yes, I agree.

Just thinking aloud (I expect similar thoughts to others): I would expect that a monotonic log's API would identify "space" in the log using either numbers for the entries, and/or would permit "global offsets" as if the log where a giant text file.

In both cases, the numbers are logical (in terms of the text content), and do not mention any "addresses"; each addressing scheme can be used to build more useful indexes, stored separately, for keyword search, or date search, etc.

(For this simple example, we would accept that those extra features aren't going to exist, and the ability to inspect the log is limited to looking at its tail, or paying more, and looking in O(N) time back an arbitrary `N` entries.)

@rossberg -- Regarding your thoughts toward a more clever (tree) representation where we store more meta data as a balanced tree of some kind.  Rather than design something new or choose something that requires re-balancing at all (this is stable memory and it would be nice to avoid that), my conception of that would be to use a cartesian tree data structure that is balanced in expectation as a it grows, and never rebalances.

- https://en.wikipedia.org/wiki/Cartesian_tree
- I've already [implemented them in Motoko, as a vessel library](https://github.com/matthewhammer/motoko-sequence/blob/e57b88cf4aa4852c7f66b9150692e256911c1425/src/Sequence.mo#L7); we used these in CanCan, though we never tested the scaling properties of them there.
- I learned about these from a [POPL 1989 paper about representations for caching and sharing data](http://matthewhammer.org/courses/csci7000-s17/readings/Pugh89.pdf); the paper calls them "level trees", rather than "cartesian trees".

It's like a balanced rope, except that by adding some extra meta data up front to each internal node, we avoid ever needing to rebalance at all.  Then, we get the time bounds that Andreas mentioned above:
- random query/update accesses from the root are O(log n) time, 
- and query/update to the end can be made in (I think?), amortized O(1) time [by using a variation of the tree that is opened like a zipper](https://arxiv.org/abs/1608.06009).